src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -116,6 +149,67 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
 
 
 /*
+ * InitTransactionStateForTask is called during executor start with the first
+ * modifying (INSERT/UPDATE/DELETE) task during a transaction. It creates the
+ * transaction participant hash, opens connections to this task's nodes, and
+ * populates the hash with those connections after sending BEGIN commands to
+ * each. If a node fails to respond, its connection is set to NULL to prevent
+ * further interaction with it during the transaction.
+ */
+static void
+InitTransactionStateForTask(Task *task)
+{
+	MemoryContext oldContext = NULL;
+	ListCell *placementCell = NULL;
+
+	xactParticipantHash = CreateXactParticipantHash();
+
+	oldContext = MemoryContextSwitchTo(TopTransactionContext);
```

Comment:
Good catch. We _were_ when I had the `shardId` list on each hash entry (via `lappend`) but now are not. Removing.